// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "google/api/annotations.proto";
// import "google/longrunning/operations.proto";
import "google/protobuf/duration.proto";
import "google/rpc/status.proto";

package google.showcase.v1alpha2;

option go_package = "github.com/googleapis/gapic-showcase/server/genproto";
option java_package = "com.google.showcase.v1alpha2";
option java_multiple_files = true;

// option (google.api.metadata) = {
//   product_uri: "https://github.com/googleapis/gapic-showcase"
// };

// This service is used showcase the four main types of rpcs - unary, server
// side streaming, client side streaming, and bidirectional streaming. This
// service also exposes methods that explicitly implement server delay, and
// paginated calls.
service Echo {
  // This service is meant to only run locally on the port 7469 (keypad digits
  // for "show").
  // option (google.api.default_host) = "localhost:7469";

  // This method simply echos the request. This method is showcases unary rpcs.
  rpc Echo(EchoRequest) returns (EchoResponse) {
    option (google.api.http) = {
      post: "/v1alpha2/echo:echo"
      body: "*"
    };
  }

  // This method split the given content into words and will pass each word back
  // through the stream. This method showcases server-side streaming rpcs.
  rpc Expand(ExpandRequest) returns (stream EchoResponse) {
    option (google.api.http) = {
      post: "/v1alpha2/echo:expand"
      body: "*"
    };
    // TODO(landrito): change this to be `fields: ["content", "error"]` once
    // github.com/dcodeIO/protobuf.js/issues/1094 has been resolved.
    // option (google.api.method_signature) = {
    //   fields: "content"
    //   fields: "error"
    // };
  }

  // This method will collect the words given to it. When the stream is closed
  // by the client, this method will return the a concatenation of the strings
  // passed to it. This method showcases client-side streaming rpcs.
  rpc Collect(stream EchoRequest) returns (EchoResponse) {
    option (google.api.http) = {
      post: "/v1alpha2/echo:collect"
      body: "*"
    };
  }

  // This method, upon receiving a request on the stream, the same content will
  // be passed  back on the stream. This method showcases bidirectional
  // streaming rpcs.
  rpc Chat(stream EchoRequest) returns (stream EchoResponse);

  // This method will wait the requested amount of and then return.
  // This method showcases how a client handles a request timing out.
  rpc Wait(WaitRequest) returns (WaitResponse) {
    option (google.api.http) = {
      post: "/v1alpha2/echo:wait"
      body: "*"
    };
  }

  // This method returns a page containing numbers specified in the request.
  // If the last number in the page is below the specified maximum, then the
  // response will include a page token indicating the next page.
  rpc Pagination(PaginationRequest) returns (PaginationResponse) {
    option (google.api.http) = {
      post: "/v1alpha2/echo:pagination"
      body: "*"
    };
    // TODO(landrito): change this to be `fields: ["max_response"]` once
    // github.com/dcodeIO/protobuf.js/issues/1094 has been resolved.
    // option (google.api.method_signature) = {
    //   fields: "max_response"
    // };
  }
}

// The request message used for the Echo, Collect and Chat methods. If content
// is set in this message then the request will succeed. If a status is
message EchoRequest {
  oneof response {
    // The content to be echoed by the server.
    string content = 1;

    // The error to be thrown by the server.
    google.rpc.Status error = 2;
  }
}

// The response message for the Echo methods.
message EchoResponse {
  // The content specified in the request.
  string content = 1;
}

// The request message for the Expand method.
message ExpandRequest {
  // The content that will be split into words and returned on the stream.
  string content = 1;

  // The error that is thrown after all words are sent on the stream.
  google.rpc.Status error = 2;
}

// The request for Wait method.
message WaitRequest {
  // The amount of time to wait before returning a response.
  google.protobuf.Duration response_delay = 1; // [(google.api.required) = true];

  oneof response {
    // The error that will be returned by the server. If this code is specified
    // to be the OK rpc code, an empty response will be returned.
    google.rpc.Status error = 2;

    // The response to be returned that will signify successful method call.
    WaitResponse success = 3;
  }
}

// The response for Wait method.
message WaitResponse {
  // This content can contain anything, the server will not depend on a value
  // here.
  string content = 1;
}

// The request for the Pagination methods.
message PaginationRequest {
  // The maximum number that will be returned in the response.
  int32 max_response = 1; // [(google.api.required) = true];

  // The amount of numbers to returned in the response.
  int32 page_size = 2;

  // The position of the page to be returned. This will be a stringified int
  // that will signifiy where to start the page from. Anything other than
  // a stringified integer within the range of 0 and the max_response will
  // cause an error to be thrown. This value is a string as opposed to a int32
  // to follow general google api practices.
  string page_token = 3;
}

// The response for the Pagination method.
message PaginationResponse {
  // An increasing list of responses starting at the value specified by the
  // page token. If the page token is empty, then this list will start at 0.
  repeated int32 responses = 1;

  // The next integer stringified.
  string next_page_token = 2;
}
